### PR TITLE
XDN Sapper example

### PR DESCRIPTION
**Pull request by tristanlee85**
https://moovweb-docs-xdn-sapper-example-staging.moovweb-edge.io/

### CODE REVIEW

**Comment by tristanlee85** [comment]
https://moovweb-docs-xdn-sapper-example-staging.moovweb-edge.io/

File: xdn-next-example/xdn/cache.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      maxAgeSeconds: 0,
+      serviceWorkerSeconds: 60 * 60 * 24,
+    },
+    edge: {
+      maxAgeSeconds: 60 * 60 * 24 * 365 * 10,
+      staleWhileRevalidateSeconds: 60 * 60 * 24,
+    },
+  })
+  proxy('api')
+  setResponseHeader('content-type', 'application/json')
```
Comment (markbrocato) [inline xdn-next-example/xdn/cache.js]: You should set this content-type header on the API app. That way you don't need to set it in each example app.

File: xdn-next-example/xdn/cache.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    browser: {
+      maxAgeSeconds: 0,
+      serviceWorkerSeconds: 60 * 60 * 24,
+    },
+    edge: {
+      maxAgeSeconds: 60 * 60 * 24 * 365 * 10,
+      staleWhileRevalidateSeconds: 60 * 60 * 24,
+    },
+  })
+  setResponseHeader('content-type', 'application/json')
```
Comment (markbrocato) [inline xdn-next-example/xdn/cache.js]: Same here

File: xdn-sapper-example/src/client.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+sapper.start({
+	target: document.querySelector('#sapper')
+}).then(() => {
+  const { serviceWorker } = navigator
+  if (serviceWorker) {
+    serviceWorker.addEventListener('message', event => {
+      if (event.data.action === 'prefetch') {
+        prefetch(event.data.url, event.data.as, event.data.options)
+      }
+    })
```
Comment (markbrocato) [inline xdn-sapper-example/src/client.js]: Why is this needed? Shouldn't @xdn/prefetch handle this automatically?

File: xdn-sapper-example/src/components/Nav.svelte [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+              <a
+                aria-current={$page.path === category.href ? 'page' : undefined}
+                href={category.href}>{category.categoryName}</a>
+            </Prefetch>
+          </li>
+        {/each}
+      </ul>
+    </div>
+  </header>
+  <!-- <div class="container">{displayBackButton && <BackButton />}</div> -->
```
Comment (markbrocato) [inline xdn-sapper-example/src/components/Nav.svelte:97]: You should remove this.

Comment (markbrocato) [inline xdn-sapper-example/src/components/Nav.svelte]:
```suggestion
```

File: xdn-sapper-example/src/routes/_error.svelte [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<script>
```
Comment (markbrocato) [inline xdn-sapper-example/src/routes/_error.svelte:1]: You should convert this file from tabs to spaces to be consistent with the rest of the sources.

File: xdn-sapper-example/src/routes/_layout.svelte [hunk 6]
```diff
@@ -0,0 +1,24 @@
+<script>
```
Comment (markbrocato) [inline xdn-sapper-example/src/routes/_layout.svelte:1]: Same: tabs => spaces

File: xdn-sapper-example/lib/cms.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
+import fetch from 'axios';
+
+const origin = 'https://moovweb-docs-xdn-examples-api-default.moovweb-edge.io'
+let apiUrl
+
+if (typeof window !== 'undefined') {
+  apiUrl = location.protocol + '//' + location.host + '/api'
```
Comment (markbrocato) [inline xdn-sapper-example/lib/cms.js]: You should always add a build number to API urls to ensure that when the browser downloads a new client, it doesn't use API responses cached in the service worker by the previous version of the app.  Next.js getServerSideProps automatically does this, but other frameworks such as Sapper (and probably Nuxt, not sure) do not.

File: xdn-sapper-example/webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
+const webpack = require('webpack');
```
Comment (markbrocato) [inline xdn-sapper-example/webpack.config.js]: Why have a rollup and webpack config?

File: xdn-sapper-example/xdn/cache.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+export const SSR_CACHE_HANDLER = ({ removeUpstreamResponseHeader, cache }) => {
+  removeUpstreamResponseHeader('cache-control')
+  cache({
+    browser: {
+      maxAgeSeconds: 0,
+    },
```
Comment (markbrocato) [inline xdn-sapper-example/xdn/cache.js]:
```suggestion
    browser: false,
```

This is a bit more readable ^^

File: xdn-next-example/xdn/cache.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
+export const SSR_CACHE_HANDLER = ({ removeUpstreamResponseHeader, cache }) => {
+  removeUpstreamResponseHeader('cache-control')
+  cache({
+    browser: {
+      maxAgeSeconds: 0,
+    },
```
Comment (markbrocato) [inline xdn-next-example/xdn/cache.js]:
```suggestion
    browser: false,
```

^^ This is a bit more readable.

File: xdn-sapper-example/webpack.config.js (hunk 8 above)
Comment (tristanlee85) [reply xdn-sapper-example/webpack.config.js]: Rollup was left as a backup example, but it is now removed.

File: xdn-sapper-example/src/client.js (hunk 3 above)
Comment (tristanlee85) [reply xdn-sapper-example/src/client.js]: No longer needed, was an unnecessary artifact from Next example

File: xdn-sapper-example/lib/cms.js (hunk 7 above)
Comment (tristanlee85) [reply xdn-sapper-example/lib/cms.js]: I added a build timestamp defined in webpack config that is set in the env to reference. Getting the Sapper build ID at build- and run-time seemed a bit more complex as build.json isn't available to import during the build.
